src/csharp/Grpc.Tools/build/_protobuf/Google.Protobuf.Tools.targets

```diff
@@ -38,6 +39,16 @@
     <AvailableItemName Include="Protobuf" />
   </ItemGroup>
 
+  <!-- Design Time Build integration for Visual Studio -->
+  <ItemGroup Condition=" '$(DisableProtobufDesignTimeBuild)' != 'true' " >
+    <!-- Add Protobuf items to Content if they are not added yet -->
+    <UnaddedProtobufContent Include="@(Protobuf)" Exclude="@(Content)" />
+    <Content Include="@(UnaddedProtobufContent)" />
```

Comment:
doesn't "Content" imply that the file will get copied to the target directory during the build? (I think at least in older .net version that used to be the case). Also there was some kind of transitivity behavior.

Reply:
Correct. There is a metadatum that controls this, CopyToOutputDirectory or similar. I'll be at my keyboard later today, but you'll quickly find the exact name, it's very commonly used. It's `true`  for Content and `false` for None by default, but can be set per-item, `<Content Include=... CTOD='false' />`. NuGet may also pack them, please check if it needs more metadata to disable this. I *think* it respects this copy to output value, but please better make sure.

This kinda borders on a hack. I'm ok with this if there is really totally clearly no other way to trigger a design time build (DTB), but I'd rather make very sure. AFAICR, previously VS did respect the SourceFilesProjectOutputGroupOutput--or did it only for inter-project dependencies, but not for DTB? In any case, it would be super clean if you could make a request to the VS guys to add a similar supported mechanism for DTB. (I do not know whether the DTB support part is in the roslyn repo on GitHub, or internal to MS--do you, btw?) I'm not totally at ease with adding stuff to Content, as this has a potential of opening a Pandora box of strange behaviors in the field: This is a very old list (it has always been there, I believe), exists in all builds, and we cannot know how people use it in the wild, and where would .proto files end up in their setup. I would certainly feel bad if someone publishes internal sources on their Web site because of this just by upgrading the Tools package.

Please rename `UnaddedProtobufContent` to `_Protobuf_TempUnaddedContent` or something in this vein. This list should not be non-`_` in any case. I remember we also agreed at some point to rename DisableProtobufDesignTimeBuild to Protobuf_... something, to go along with the existing "namespace" prefix convention.

Tangentially, I would not bet my life on this statement, but I'm nearly sure `<Content Include='@(Something) Exclude='@(Content)' />` will do the trick w/o an intermediate list, i.e. the list augmentation is treated generally as `ListName += (Include - Exclude)`. I'll check later today, I'm myself wondering if this is so! :)

Reply:
Thanks for the feedback. I wasn't thrilled with these changes which seemed to be hacky to me as well but it turns out there was a bug in the project system and I was just working around it. I'll test these changes with 16.1 and if it works I'll mark the PR as ready to review.

Reply:
Great that we can do it without a hack!

Reply:
Just FYI, probably not relevant any more as the issue has been fixed in dotnetbuild, but, as I suspected, an intermediate list (like UnaddedProtobufContent here) is not required.

test.xml
```xml
<Project DefaultTargets="Build">
  <ItemGroup>
    <Proto Include="a;b;c;1" />
    <Content Include="a;1;2;3" />
    
    <Content Include="@(Proto)" Exclude="@(Content)" />
  </ItemGroup>
  
  <Target Name="Build">
    <Message Importance="high" Text="@(Content)" />
  </Target>
</Project>
```
Result:
```
c:\temp>msbuild test.xml
Microsoft (R) Build Engine version 15.9.21+g9802d43bc3 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 19-04-09 16:24:13.
Project "c:\temp\test.xml" on node 1 (default targets).
Build:
  a;1;2;3;b;c
Done Building Project "c:\temp\test.xml" (default targets).
```

So the general pattern `<X Include="@(Y)" Exclude="@(X) />` adds to the end of X only those items in Y that do not have corresponding identically named items in X already.